File: georocket-server/src/test/java/io/georocket/storage/mongodb/MongoDBStoreTest.java
package io.georocket.storage.mongodb;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.BeforeClass;

import com.google.common.collect.Iterables;
import com.mongodb.MongoClient;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.gridfs.GridFSBucket;
import com.mongodb.client.gridfs.GridFSBuckets;
import com.mongodb.client.gridfs.GridFSFindIterable;
import com.mongodb.client.gridfs.model.GridFSFile;

import io.georocket.constants.ConfigConstants;
import io.georocket.storage.StorageTest;
import io.georocket.storage.Store;
import io.georocket.util.PathUtils;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.VertxOptions;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.RunTestOnContext;

/**
 * Test {@link MongoDBStore}
 * @author Andrej Sajenko
 */
public class MongoDBStoreTest extends StorageTest {
  private static long MAX_WORKER_EXECUTION_TIME = 30 * 60 * 1000;
  
  private static MongoDBTestConnector mongoConnector;

  /**
   * Default constructor
   */
  public MongoDBStoreTest() {
    super.rule = new RunTestOnContext(new VertxOptions()
        .setMaxWorkerExecuteTime(MAX_WORKER_EXECUTION_TIME));
  }

  /**
   * Set up test dependencies.
   * @throws IOException if the MongoDB instance could not be started
   */
  @BeforeClass
  public static void setUpClass() throws IOException {
    mongoConnector = new MongoDBTestConnector();
  }

  /**
   * Uninitialize tests
   */
  @AfterClass
  public static void tearDownClass() {
    mongoConnector.stop();
    mongoConnector = null;
  }
  
  /**
   * Uninitialize tests
   */
  @After
  public void tearDown() {
    try (MongoClient client = new MongoClient(mongoConnector.serverAddress)) {
      MongoDatabase db = client.getDatabase(MongoDBTestConnector.MONGODB_DBNAME);
      db.drop();
    }
  }

  private void configureVertx(Vertx vertx) {
    JsonObject config = vertx.getOrCreateContext().config();

    config.put(ConfigConstants.STORAGE_MONGODB_CONNECTION_STRING,
        "mongodb://" + mongoConnector.serverAddress.getHost() + ":" +
        mongoConnector.serverAddress.getPort());
    config.put(ConfigConstants.STORAGE_MONGODB_DATABASE,
        MongoDBTestConnector.MONGODB_DBNAME);
  }

  @Override
  protected Store createStore(Vertx vertx) {
    configureVertx(vertx);
    return new MongoDBStore(vertx);
  }

  @Override
  protected void prepareData(TestContext context, Vertx vertx, String path,
      Handler<AsyncResult<String>> handler) {
    String filename = PathUtils.join(path, ID);
    vertx.<String>executeBlocking(f -> {
      try (MongoClient client = new MongoClient(mongoConnector.serverAddress)) {
        MongoDatabase db = client.getDatabase(MongoDBTestConnector.MONGODB_DBNAME);
        GridFSBucket gridFS = GridFSBuckets.create(db);
        byte[] contents = CHUNK_CONTENT.getBytes(StandardCharsets.UTF_8);
        gridFS.uploadFromStream(filename, new ByteArrayInputStream(contents));
        f.complete(filename);
      }
    }, handler);
  }

  @Override
  protected void validateAfterStoreAdd(TestContext context, Vertx vertx,
      String path, Handler<AsyncResult<Void>> handler) {
    vertx.executeBlocking(f -> {
      try (MongoClient client = new MongoClient(mongoConnector.serverAddress)) {
        MongoDatabase db = client.getDatabase(MongoDBTestConnector.MONGODB_DBNAME);
        GridFSBucket gridFS = GridFSBuckets.create(db);
  
        GridFSFindIterable files = gridFS.find();
  
        GridFSFile file = files.first();
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        gridFS.downloadToStream(file.getFilename(), baos);
        String contents = new String(baos.toByteArray(), StandardCharsets.UTF_8);
        context.assertEquals(CHUNK_CONTENT, contents);
      }
      f.complete();
    }, handler);
  }

  @Override
  protected void validateAfterStoreDelete(TestContext context, Vertx vertx,
      String path, Handler<AsyncResult<Void>> handler) {
    vertx.executeBlocking(f -> {
      try (MongoClient client = new MongoClient(mongoConnector.serverAddress)) {
        MongoDatabase db = client.getDatabase(MongoDBTestConnector.MONGODB_DBNAME);
        GridFSBucket gridFS = GridFSBuckets.create(db);
  
        GridFSFindIterable files = gridFS.find();
        context.assertTrue(Iterables.isEmpty(files));
      }
      f.complete();
    }, handler);
  }
}

File: georocket-server/src/test/kotlin/io/georocket/storage/mongodb/MongoDBStoreTest.kt
package io.georocket.storage.mongodb

import com.mongodb.reactivestreams.client.MongoClients
import com.mongodb.reactivestreams.client.gridfs.GridFSBuckets
import io.georocket.storage.StorageTest
import io.georocket.storage.Store
import io.georocket.util.PathUtils
import io.vertx.core.Vertx
import io.vertx.junit5.VertxTestContext
import kotlinx.coroutines.reactive.awaitFirst
import kotlinx.coroutines.reactive.awaitSingle
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeAll
import reactor.core.publisher.Mono
import java.nio.ByteBuffer

/**
 * Test [MongoDBStore]
 * @author Andrej Sajenko
 */
class MongoDBStoreTest : StorageTest() {
  companion object {
    private lateinit var mongoConnector: MongoDBTestConnector

    /**
     * Set up test dependencies
     */
    @BeforeAll
    @JvmStatic
    fun setUpClass() {
      mongoConnector = MongoDBTestConnector()
    }

    /**
     * Uninitialize tests
     */
    @AfterAll
    @JvmStatic
    fun tearDownClass() {
      mongoConnector.stop()
    }
  }

  /**
   * Uninitialize tests
   */
  @AfterEach
  fun tearDown() {
    return MongoClients.create(mongoConnector.connectionString).use { client ->
      val db = client.getDatabase(mongoConnector.connectionString.database)
      db.drop()
    }
  }

  override fun createStore(vertx: Vertx): Store {
    return MongoDBStore(vertx, mongoConnector.connectionString.toString())
  }

  override suspend fun prepareData(ctx: VertxTestContext, vertx: Vertx,
      path: String?): String {
    val filename = PathUtils.join(path, ID)
    return MongoClients.create(mongoConnector.connectionString).use { client ->
      val db = client.getDatabase(mongoConnector.connectionString.database)
      val gridFS = GridFSBuckets.create(db)
      val contents = ByteBuffer.wrap(CHUNK_CONTENT.toByteArray())
      gridFS.uploadFromPublisher(filename, Mono.just(contents)).awaitSingle()
      filename
    }
  }

  override suspend fun validateAfterStoreAdd(ctx: VertxTestContext, vertx: Vertx,
      path: String?) {
    MongoClients.create(mongoConnector.connectionString).use { client ->
      val db = client.getDatabase(mongoConnector.connectionString.database)
      val gridFS = GridFSBuckets.create(db)
      val files = gridFS.find()
      val file = files.awaitFirst()
      val publisher = gridFS.downloadToPublisher(file.filename)
      val bytebuf = publisher.awaitSingle()
      val contents = String(bytebuf.array())
      assertThat(contents).isEqualTo(CHUNK_CONTENT)
    }
  }

  override suspend fun validateAfterStoreDelete(ctx: VertxTestContext,
      vertx: Vertx, path: String) {
    MongoClients.create(mongoConnector.connectionString).use { client ->
      val db = client.getDatabase(mongoConnector.connectionString.database)
      val filesCollection = db.getCollection("fs.files")
      val count = filesCollection.countDocuments().awaitSingle()
      assertThat(count).isEqualTo(0)
    }
  }
}
